extensions/positron-python/python_files/posit/positron/tests/test_positron_jedilsp.py

```diff
@@ -1042,3 +1048,37 @@ def test_positron_rename(
     text_document_edit = workspace_edit.document_changes[0]
     assert isinstance(text_document_edit, TextDocumentEdit)
     assert text_document_edit.edits == expected_text_edits
+
+
+def test_cancel_request_immediately() -> None:
+    # Test our workaround to a pygls performance issue where the server still executes requests
+    # even if they're immediately cancelled.
+    # See: https://github.com/openlawlibrary/pygls/issues/517.
+    server = create_server()
+    protocol = server.lsp
+
+    # Register a textDocument/hover handler and track whether it executes.
+    did_hover_executed = False
+
+    @server.feature(TEXT_DOCUMENT_HOVER)
+    def _hover(_server: PositronJediLanguageServer, _params: TextDocumentPositionParams):
+        nonlocal did_hover_executed
+        did_hover_executed = True
+
+    # Helper to encode LSP messages, adapted from `pygls.json_rpc.JsonRPCProtocol._send_data`.
+    def encode(message):
+        body = json.dumps(message, default=protocol._serialize_message)  # noqa: SLF001
+        header = f"Content-Length: {len(body)}\r\n\r\n"
+        return (header + body).encode(protocol.CHARSET)
+
+    # Send a hover request and immediately cancel it in the same payload.
+    hover_request = TextDocumentHoverRequest(
+        0, HoverParams(TextDocumentIdentifier(""), Position(0, 0))
+    )
+    cancel_request = CancelRequestNotification(CancelParams(hover_request.id))
+    data = encode(hover_request) + encode(cancel_request)
+
+    # Call `_data_received` instead of `data_received` so that errors are raised.
+    protocol._data_received(data)  # noqa: SLF001
+
+    assert not did_hover_executed
```

Comment:
Like you said, not sure how to test this by hovering locally, but this unit test looks good!